Fix #153/#154 : MultipleChoice/ItemSelection Interaction View

veena14cs opened the pull request:
##  Explanation

This PR is duplicate of [#184](https://github.com/oppia/oppia-android/pull/184). It contains UI structure for Multiple choice and item selection interactions, and fixes #153 and #154.

Reference Design Doc: https://docs.google.com/document/d/1HFwDFcqenAXepbYr5rfawGHvkeBZpYvNMs228AWLn24/edit?usp=sharing

veena14cs committed:
added interactions

veena14cs committed:
Update ContentInteractionFragmentPresenter.kt

veena14cs committed:
Update ContentInteractionFragmentPresenter.kt

rt4914 commented on app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragment.kt:
```diff
@@ -0,0 +1,26 @@
+import org.oppia.app.player.content_interaction.ContentInteractionFragmentPresenter
+import javax.inject.Inject
+
+/** Fragment displays single/multiplechoice input interaction. */
```
Do not you `single` or `singlechoice` or `singlechoiceinput` anywhere, in our discussion that was just for communication.
If you are using `SingleChoiceInput` -> that by default developer will understand that in this radio buttons should be visible and leaner can select only one correct answer. This then drives the assumption that `MultipleChoiceInput` will show checkboxes and learner can select multiple answers.

This is incorrect in terms of definition as well as oppia-web codebase.

Use `ItemSelectionInput` or `ItemSelectionViewInteraction` whichever fits the name, this means that learner can select multiple answers at a time in backend code.
Use `MultipleChoiceInput` or `MultipleChoiceViewInteraction`, this means that learner can select one answer among multiple options, as per backend code.

veena14cs replied:
this file is been removed

rt4914 commented on app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragment.kt:
```diff
@@ -0,0 +1,26 @@
+
+/** Fragment displays single/multiplechoice input interaction. */
+class ContentInteractionFragment : InjectableFragment() {
+
```
Nit: Remove this empty line.

veena14cs replied:
File is removed

rt4914 commented on app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt:
```diff
@@ -0,0 +1,147 @@
+class ContentInteractionFragmentPresenter @Inject constructor(
+  @ApplicationContext private val context: Context,
+  private val fragment: Fragment
+//  private val interactionInstanceId: String, val gaeCustomizationArgs: Map<String, GaeCustomizationArgs>
```
Remove this comment and unused variable

veena14cs replied:
removed

rt4914 commented on app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt:
```diff
@@ -0,0 +1,147 @@
+
+    showInputInteractions()
+
+    return binding.root
```
Remove extra empty spaces.

rt4914 commented on app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt:
```diff
@@ -0,0 +1,147 @@
+  }
+
+  private fun showInputInteractions() {
+
```
Remove extra empty space

veena14cs replied:
removed extra spaces

rt4914 commented on app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt:
```diff
@@ -0,0 +1,147 @@
+
+  private fun showInputInteractions() {
+
+    //Todo(Veena): Remove static initialization and use values from constructor for gaeCustomizationArgsMap and interactionInstanceId
```
Space just before TODO.

This should be 
`// Todo(Veena): Remove static ...`

rt4914 commented on app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt:
```diff
@@ -0,0 +1,147 @@
+    return binding.root
+  }
+
+  private fun showInputInteractions() {
```
It would be good, if your method accepts whatever parameters it needs and then create the UI. So that, when we work on final implementation we can directly use your method.

rt4914 commented on app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt:
```diff
@@ -0,0 +1,147 @@
+
+  fun addCheckbox(optionsArray: Array<String>) {
+    for (row in 0..0) {
+      val ll = LinearLayout(context)
```
Rather than using `ll` using proper variable names like `checkboxLinearLayout` or `rootLinearLayout` or simply `linearLayout`

rt4914 commented on app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt:
```diff
@@ -0,0 +1,147 @@
+      ll.orientation = LinearLayout.VERTICAL
+
+      for (i in 0..optionsArray.size - 1) {
+        val cb = CheckBox(context)
```
Change variable name as per above comment.

rt4914 commented on app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt:
```diff
@@ -0,0 +1,147 @@
+
+  fun addRadioButtons(optionsArray: Array<String>) {
+    for (row in 0..0) {
+      val rg = RadioGroup(context)
```
Nit: Variable name

rt4914 commented on app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt:
```diff
@@ -0,0 +1,147 @@
+      rg.orientation = LinearLayout.VERTICAL
+
+      for (i in 0..optionsArray.size - 1) {
+        val rdbtn = RadioButton(context)
```
Nit: Variable name

rt4914 commented on app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt:
```diff
@@ -0,0 +1,147 @@
+      for (i in 0..optionsArray.size - 1) {
+        val rdbtn = RadioButton(context)
+        rdbtn.id = View.generateViewId()
+        rdbtn.text = convertHtmlToString(optionsArray[i], rdbtn).toString()// + rdbtn.id
```
Remove comment if not needed

rt4914 commented on app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt:
```diff
@@ -0,0 +1,147 @@
+                return
+              }
+            }
+
```
Remove this extra space

rt4914 commented on app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt:
```diff
@@ -0,0 +1,147 @@
+  }
+
+  fun convertHtmlToString(rawResponse: Any?, rdbtn: View): Spanned {
+
```
Remove this extra space

rt4914 commented on app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt:
```diff
@@ -0,0 +1,147 @@
+    var CUSTOM_TAG = "oppia-noninteractive-image"
+    var HTML_TAG = "img"
+    var CUSTOM_ATTRIBUTE = "filepath-with-value"
+    var HTML_ATTRIBUTE = "src"
```
I think all these variables can be converted to val. But better you define all these as `const val` before above the entire-class-code

rt4914 commented on app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt:
```diff
@@ -0,0 +1,147 @@
+    var HTML_ATTRIBUTE = "src"
+
+    if (htmlContent!!.contains(CUSTOM_TAG)) {
+
```
Remove this extra line.

rt4914 commented on utility/src/main/java/org/oppia/util/data/URlImageParser.kt:
```diff
@@ -0,0 +1,67 @@
+import java.net.URL
+
+/** UrlImage Parser for android TextView to load Html Image tag */
+
```
Remove this line. Also add `.` in the comment

`/** UrlImage Parser for android TextView to load Html Image tag. */`

rt4914 commented on utility/src/main/java/org/oppia/util/data/URlImageParser.kt:
```diff
@@ -0,0 +1,67 @@
+    override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
+
+      drawable = BitmapDrawable(context.resources, resource)
+
```
Nit: Remove extra space

veena14cs replied:
removed spaces

rt4914 commented on utility/src/main/java/org/oppia/util/data/URlImageParser.kt:
```diff
@@ -0,0 +1,67 @@
+    internal var drawable: Drawable? = null
+
+    override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
+
```
Nit: Remove extra space

veena14cs replied:
removed extra spaces

rt4914 commented on utility/src/main/java/org/oppia/util/data/URlImageParser.kt:
```diff
@@ -0,0 +1,67 @@
+        val hh = (drawable as BitmapDrawable).intrinsicHeight
+        val ww = (drawable as BitmapDrawable).intrinsicWidth
+        val newHeight = hh * w / ww
+        val rect = Rect(0, 0, w, newHeight)
```
Name variables correctly

veena14cs replied:
variable names corrected

rt4914 commented on utility/src/main/java/org/oppia/util/data/URlImageParser.kt:
```diff
@@ -0,0 +1,67 @@
+        tvContents.text = tvContents.text
+        tvContents.invalidate()
+      }
+
```
Nit: Remove extra space

veena14cs replied:
removed blank spaces

rt4914 commented on utility/src/main/java/org/oppia/util/data/URlImageParser.kt:
```diff
@@ -0,0 +1,67 @@
+        drawable!!.draw(canvas)
+    }
+  }
+
```
Nit: Remove extra space

veena14cs replied:
removed spaces

rt4914 commented:
@veena14cs 

Your code for checkbox and radio button works correct. Great job.

I have left some comments which you can resolve.

Also, you are using `ContentInteractionFragment` just for showing this demo and it actually serves no other purpose. And we know that we have to use these two views in the `StateFragment`, so I suggest you remove `ContentInteractionFragment` and shift your implementation in `StateFragment`.

Also, create a customview which handles your two views, checkout @nikitamarysolomanpvt PR on FractionInput, she has created a separate `CustomView` rather than handling everything in parent fragment.

Write test-cases for this code.

rt4914 commented:
@veena14cs  your branch name contains '()' brackets, do not add brackets to a branch name, it is wrong convention. Also, if the branch name contains `()` then switching branches from terminal is not possible directly.

rt4914 reviewed (commented):
PTAL

rt4914 commented on app/src/main/res/layout/content_interaction_fragment.xml:
```diff
@@ -0,0 +1,14 @@
+        android:id="@+id/radioGroup"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+      android:padding="15dp">
```
Indentation issue.

veena14cs replied:
this file is been removed

veena14cs committed:
updated urlImageParser file  name

veena14cs committed:
created customviews. moved implementation to statefragment

veena14cs commented:
Moved Interaction implementation to statefragment.

On Fri, Sep 27, 2019 at 8:52 AM Rajat Talesra <notifications@github.com>
wrote:

> @veena14cs <https://github.com/veena14cs>
>
> Your code for checkbox and radio button works correct. Great job.
>
> I have left some comments which you can resolve.
>
> Also, you are using ContentInteractionFragment just for showing this demo
> and it actually serves no other purpose. And we know that we have to use
> these two views in the StateFragment, so I suggest you remove
> ContentInteractionFragment and shift your implementation in StateFragment.
>
> Write test-cases for this code.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia-android/pull/188?email_source=notifications&email_token=AD4LXZGJTKBOLDDPRXRFAF3QLV4ABA5CNFSM4I2WFUPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7XTKUA#issuecomment-535770448>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD4LXZB4UFTSLCINF46ZV53QLV4ABANCNFSM4I2WFUPA>
> .
>

veena14cs commented:
used the terms MultipleChoiceViewInteraction and ItemSelectionViewInteraction
for interactions.
But these terms are really creating confusion though it is according to
webcode base, in my opinion we have to change these terms according to
their definations

On Fri, Sep 27, 2019 at 8:55 AM Rajat Talesra <notifications@github.com>
wrote:

> *@rt4914* commented on this pull request.
>
> PTAL
> ------------------------------
>
> In
> app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragment.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328898824>:
>
> > @@ -0,0 +1,26 @@
> +package org.oppia.app.player.content_interaction
> +
> +import android.content.Context
> +import android.os.Bundle
> +import android.view.LayoutInflater
> +import android.view.View
> +import android.view.ViewGroup
> +import org.oppia.app.fragment.InjectableFragment
> +import org.oppia.app.player.content_interaction.ContentInteractionFragmentPresenter
> +import javax.inject.Inject
> +
> +/** Fragment displays single/multiplechoice input interaction. */
>
> Do not you single or singlechoice or singlechoiceinput anywhere, in our
> discussion that was just for communication.
> If you are using SingleChoiceInput -> that by default developer will
> understand that in this radio buttons should be visible and leaner can
> select only one correct answer. This then drives the assumption that
> MultipleChoiceInput will show checkboxes and learner can select multiple
> answers.
>
> This is incorrect in terms of definition as well as oppia-web codebase.
>
> Use ItemSelectionInput or ItemSelectionViewInteraction whichever fits the
> name, this means that learner can select multiple answers at a time in
> backend code.
> Use MultipleChoiceInput or MultipleChoiceViewInteraction, this means that
> learner can select one answer among multiple options, as per backend code.
> ------------------------------
>
> In
> app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragment.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328898899>:
>
> > @@ -0,0 +1,26 @@
> +package org.oppia.app.player.content_interaction
> +
> +import android.content.Context
> +import android.os.Bundle
> +import android.view.LayoutInflater
> +import android.view.View
> +import android.view.ViewGroup
> +import org.oppia.app.fragment.InjectableFragment
> +import org.oppia.app.player.content_interaction.ContentInteractionFragmentPresenter
> +import javax.inject.Inject
> +
> +/** Fragment displays single/multiplechoice input interaction. */
> +class ContentInteractionFragment : InjectableFragment() {
> +
>
> Nit: Remove this empty line.
> ------------------------------
>
> In
> app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328898964>:
>
> > +import org.oppia.app.databinding.ContentInteractionFragmentBinding
> +import javax.inject.Inject
> +import android.widget.RadioButton
> +import android.widget.LinearLayout
> +import android.widget.RadioGroup
> +import android.widget.TextView
> +import android.widget.Toast
> +import kotlinx.android.synthetic.main.content_interaction_fragment.view.*
> +import org.oppia.util.data.UrlImageParser
> +import android.widget.CompoundButton
> +
> +/** Presenter for [ContentInteractionFragment]. */
> +class ContentInteractionFragmentPresenter @Inject constructor(
> +  @ApplicationContext private val context: Context,
> +  private val fragment: Fragment
> +//  private val interactionInstanceId: String, val gaeCustomizationArgs: Map<String, GaeCustomizationArgs>
>
> Remove this comment and unused variable
> ------------------------------
>
> In
> app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328899007>:
>
> > +/** Presenter for [ContentInteractionFragment]. */
> +class ContentInteractionFragmentPresenter @Inject constructor(
> +  @ApplicationContext private val context: Context,
> +  private val fragment: Fragment
> +//  private val interactionInstanceId: String, val gaeCustomizationArgs: Map<String, GaeCustomizationArgs>
> +) {
> +
> +  private lateinit var binding: ContentInteractionFragmentBinding
> +
> +  fun handleCreateView(inflater: LayoutInflater, container: ViewGroup?): View? {
> +
> +    binding = ContentInteractionFragmentBinding.inflate(inflater, container, /* attachToRoot= */ false)
> +
> +    showInputInteractions()
> +
> +    return binding.root
>
> Remove extra empty spaces.
> ------------------------------
>
> In
> app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328899033>:
>
> > +//  private val interactionInstanceId: String, val gaeCustomizationArgs: Map<String, GaeCustomizationArgs>
> +) {
> +
> +  private lateinit var binding: ContentInteractionFragmentBinding
> +
> +  fun handleCreateView(inflater: LayoutInflater, container: ViewGroup?): View? {
> +
> +    binding = ContentInteractionFragmentBinding.inflate(inflater, container, /* attachToRoot= */ false)
> +
> +    showInputInteractions()
> +
> +    return binding.root
> +  }
> +
> +  private fun showInputInteractions() {
> +
>
> Remove extra empty space
> ------------------------------
>
> In
> app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328899141>:
>
> > +) {
> +
> +  private lateinit var binding: ContentInteractionFragmentBinding
> +
> +  fun handleCreateView(inflater: LayoutInflater, container: ViewGroup?): View? {
> +
> +    binding = ContentInteractionFragmentBinding.inflate(inflater, container, /* attachToRoot= */ false)
> +
> +    showInputInteractions()
> +
> +    return binding.root
> +  }
> +
> +  private fun showInputInteractions() {
> +
> +    //Todo(Veena): Remove static initialization and use values from constructor for gaeCustomizationArgsMap and interactionInstanceId
>
> Space just before TODO.
>
> This should be
> // Todo(Veena): Remove static ...
> ------------------------------
>
> In
> app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328899294>:
>
> > +  private val fragment: Fragment
> +//  private val interactionInstanceId: String, val gaeCustomizationArgs: Map<String, GaeCustomizationArgs>
> +) {
> +
> +  private lateinit var binding: ContentInteractionFragmentBinding
> +
> +  fun handleCreateView(inflater: LayoutInflater, container: ViewGroup?): View? {
> +
> +    binding = ContentInteractionFragmentBinding.inflate(inflater, container, /* attachToRoot= */ false)
> +
> +    showInputInteractions()
> +
> +    return binding.root
> +  }
> +
> +  private fun showInputInteractions() {
>
> It would be good, if your method accepts whatever parameters it needs and
> then create the UI. So that, when we work on final implementation we can
> directly use your method.
> ------------------------------
>
> In
> app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328899544>:
>
> > +    if (interactionInstanceId.equals("MultipleChoiceInput")) {
> +      val gaeCustomArgsInString: String = gaeCustomizationArgs.toString().replace("[", "").replace("]", "")
> +      var items = gaeCustomArgsInString.split(",").toTypedArray()
> +      addRadioButtons(items)
> +    } else if (interactionInstanceId.equals("ItemSelectionInput") || interactionInstanceId.equals("SingleChoiceInput")) {
> +      val gaeCustomArgsInString: String = gaeCustomizationArgs.toString().replace("[", "").replace("]", "")
> +      var items = gaeCustomArgsInString.split(",").toTypedArray()
> +      addCheckbox(items)
> +    } else {
> +      //Do no show any view
> +    }
> +  }
> +
> +  fun addCheckbox(optionsArray: Array<String>) {
> +    for (row in 0..0) {
> +      val ll = LinearLayout(context)
>
> Rather than using ll using proper variable names like checkboxLinearLayout
> or rootLinearLayout or simply linearLayout
> ------------------------------
>
> In
> app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328899590>:
>
> > +    } else if (interactionInstanceId.equals("ItemSelectionInput") || interactionInstanceId.equals("SingleChoiceInput")) {
> +      val gaeCustomArgsInString: String = gaeCustomizationArgs.toString().replace("[", "").replace("]", "")
> +      var items = gaeCustomArgsInString.split(",").toTypedArray()
> +      addCheckbox(items)
> +    } else {
> +      //Do no show any view
> +    }
> +  }
> +
> +  fun addCheckbox(optionsArray: Array<String>) {
> +    for (row in 0..0) {
> +      val ll = LinearLayout(context)
> +      ll.orientation = LinearLayout.VERTICAL
> +
> +      for (i in 0..optionsArray.size - 1) {
> +        val cb = CheckBox(context)
>
> Change variable name as per above comment.
> ------------------------------
>
> In
> app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328899704>:
>
> > +        cb.text = convertHtmlToString(optionsArray[i], cb).toString()
> +
> +        cb.setOnCheckedChangeListener(CompoundButton.OnCheckedChangeListener { buttonView, isChecked ->
> +          val msg = "You have " + (if (isChecked) "checked" else "unchecked") + " this Check it Checkbox."
> +          Toast.makeText(context, msg, Toast.LENGTH_SHORT).show()
> +        })
> +
> +        ll.addView(cb)
> +      }
> +      binding.root.radioGroup.addView(ll)
> +    }
> +  }
> +
> +  fun addRadioButtons(optionsArray: Array<String>) {
> +    for (row in 0..0) {
> +      val rg = RadioGroup(context)
>
> Nit: Variable name
> ------------------------------
>
> In
> app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328899715>:
>
> > +          Toast.makeText(context, msg, Toast.LENGTH_SHORT).show()
> +        })
> +
> +        ll.addView(cb)
> +      }
> +      binding.root.radioGroup.addView(ll)
> +    }
> +  }
> +
> +  fun addRadioButtons(optionsArray: Array<String>) {
> +    for (row in 0..0) {
> +      val rg = RadioGroup(context)
> +      rg.orientation = LinearLayout.VERTICAL
> +
> +      for (i in 0..optionsArray.size - 1) {
> +        val rdbtn = RadioButton(context)
>
> Nit: Variable name
> ------------------------------
>
> In
> app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328899742>:
>
> > +
> +        ll.addView(cb)
> +      }
> +      binding.root.radioGroup.addView(ll)
> +    }
> +  }
> +
> +  fun addRadioButtons(optionsArray: Array<String>) {
> +    for (row in 0..0) {
> +      val rg = RadioGroup(context)
> +      rg.orientation = LinearLayout.VERTICAL
> +
> +      for (i in 0..optionsArray.size - 1) {
> +        val rdbtn = RadioButton(context)
> +        rdbtn.id = View.generateViewId()
> +        rdbtn.text = convertHtmlToString(optionsArray[i], rdbtn).toString()// + rdbtn.id
>
> Remove comment if not needed
> ------------------------------
>
> In
> app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328899757>:
>
> > +        val rdbtn = RadioButton(context)
> +        rdbtn.id = View.generateViewId()
> +        rdbtn.text = convertHtmlToString(optionsArray[i], rdbtn).toString()// + rdbtn.id
> +        rdbtn.gravity = (Gravity.RIGHT); (Gravity.CENTER)
> +
> +        rg.setOnCheckedChangeListener(object : RadioGroup.OnCheckedChangeListener {
> +          override fun onCheckedChanged(group: RadioGroup, checkedId: Int) {
> +            for (i in 0 until group.childCount) {
> +              val btn = group.getChildAt(i) as RadioButton
> +              if (btn.id == checkedId) {
> +                val text = btn.text
> +                Toast.makeText(context, "" + text, Toast.LENGTH_LONG).show()
> +                return
> +              }
> +            }
> +
>
> Remove this extra space
> ------------------------------
>
> In
> app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328899776>:
>
> > +                val text = btn.text
> +                Toast.makeText(context, "" + text, Toast.LENGTH_LONG).show()
> +                return
> +              }
> +            }
> +
> +          }
> +        })
> +        rg.addView(rdbtn)
> +      }
> +      binding.root.radioGroup.addView(rg)
> +    }
> +  }
> +
> +  fun convertHtmlToString(rawResponse: Any?, rdbtn: View): Spanned {
> +
>
> Remove this extra space
> ------------------------------
>
> In
> app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328899898>:
>
> > +          }
> +        })
> +        rg.addView(rdbtn)
> +      }
> +      binding.root.radioGroup.addView(rg)
> +    }
> +  }
> +
> +  fun convertHtmlToString(rawResponse: Any?, rdbtn: View): Spanned {
> +
> +    var htmlContent = rawResponse as String;
> +    var result: Spanned
> +    var CUSTOM_TAG = "oppia-noninteractive-image"
> +    var HTML_TAG = "img"
> +    var CUSTOM_ATTRIBUTE = "filepath-with-value"
> +    var HTML_ATTRIBUTE = "src"
>
> I think all these variables can be converted to val. But better you define
> all these as const val before above the entire-class-code
> ------------------------------
>
> In
> app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328899936>:
>
> > +      }
> +      binding.root.radioGroup.addView(rg)
> +    }
> +  }
> +
> +  fun convertHtmlToString(rawResponse: Any?, rdbtn: View): Spanned {
> +
> +    var htmlContent = rawResponse as String;
> +    var result: Spanned
> +    var CUSTOM_TAG = "oppia-noninteractive-image"
> +    var HTML_TAG = "img"
> +    var CUSTOM_ATTRIBUTE = "filepath-with-value"
> +    var HTML_ATTRIBUTE = "src"
> +
> +    if (htmlContent!!.contains(CUSTOM_TAG)) {
> +
>
> Remove this extra line.
> ------------------------------
>
> In utility/src/main/java/org/oppia/util/data/URlImageParser.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328900054>:
>
> > +import android.content.Context
> +import android.graphics.Bitmap
> +import android.graphics.Canvas
> +import android.graphics.Rect
> +import android.graphics.drawable.BitmapDrawable
> +import android.graphics.drawable.Drawable
> +import android.graphics.drawable.LevelListDrawable
> +import android.text.Html
> +import android.widget.TextView
> +import com.bumptech.glide.Glide
> +import com.bumptech.glide.request.target.SimpleTarget
> +import com.bumptech.glide.request.transition.Transition
> +import java.net.URL
> +
> +/** UrlImage Parser for android TextView to load Html Image tag */
> +
>
> Remove this line. Also add . in the comment
>
> /** UrlImage Parser for android TextView to load Html Image tag. */
> ------------------------------
>
> In utility/src/main/java/org/oppia/util/data/URlImageParser.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328900117>:
>
> > +  override fun getDrawable(urlString: String): Drawable {
> +    val urlDrawable = UrlDrawable()
> +    val load = Glide.with(context).asBitmap().load(URL(urlString))
> +    val target = BitmapTarget(urlDrawable)
> +    targets?.add(target)
> +    load.into(target)
> +    return urlDrawable
> +  }
> +
> +  inner class BitmapTarget(private val urlDrawable: UrlDrawable) : SimpleTarget<Bitmap>() {
> +    internal var drawable: Drawable? = null
> +
> +    override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
> +
> +      drawable = BitmapDrawable(context.resources, resource)
> +
>
> Nit: Remove extra space
> ------------------------------
>
> In utility/src/main/java/org/oppia/util/data/URlImageParser.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328900125>:
>
> > +   * @return
> +   */
> +  override fun getDrawable(urlString: String): Drawable {
> +    val urlDrawable = UrlDrawable()
> +    val load = Glide.with(context).asBitmap().load(URL(urlString))
> +    val target = BitmapTarget(urlDrawable)
> +    targets?.add(target)
> +    load.into(target)
> +    return urlDrawable
> +  }
> +
> +  inner class BitmapTarget(private val urlDrawable: UrlDrawable) : SimpleTarget<Bitmap>() {
> +    internal var drawable: Drawable? = null
> +
> +    override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
> +
>
> Nit: Remove extra space
> ------------------------------
>
> In utility/src/main/java/org/oppia/util/data/URlImageParser.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328900186>:
>
> > +    return urlDrawable
> +  }
> +
> +  inner class BitmapTarget(private val urlDrawable: UrlDrawable) : SimpleTarget<Bitmap>() {
> +    internal var drawable: Drawable? = null
> +
> +    override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
> +
> +      drawable = BitmapDrawable(context.resources, resource)
> +
> +      tvContents.post {
> +        val w = tvContents.width
> +        val hh = (drawable as BitmapDrawable).intrinsicHeight
> +        val ww = (drawable as BitmapDrawable).intrinsicWidth
> +        val newHeight = hh * w / ww
> +        val rect = Rect(0, 0, w, newHeight)
>
> Name variables correctly
> ------------------------------
>
> In utility/src/main/java/org/oppia/util/data/URlImageParser.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328900193>:
>
> > +
> +      drawable = BitmapDrawable(context.resources, resource)
> +
> +      tvContents.post {
> +        val w = tvContents.width
> +        val hh = (drawable as BitmapDrawable).intrinsicHeight
> +        val ww = (drawable as BitmapDrawable).intrinsicWidth
> +        val newHeight = hh * w / ww
> +        val rect = Rect(0, 0, w, newHeight)
> +        (drawable as BitmapDrawable).bounds = rect
> +        urlDrawable.bounds = rect
> +        urlDrawable.drawable = drawable
> +        tvContents.text = tvContents.text
> +        tvContents.invalidate()
> +      }
> +
>
> Nit: Remove extra space
> ------------------------------
>
> In utility/src/main/java/org/oppia/util/data/URlImageParser.kt
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328900200>:
>
> > +        urlDrawable.drawable = drawable
> +        tvContents.text = tvContents.text
> +        tvContents.invalidate()
> +      }
> +
> +    }
> +  }
> +
> +  inner class UrlDrawable : BitmapDrawable() {
> +    var drawable: Drawable? = null
> +    override fun draw(canvas: Canvas) {
> +      if (drawable != null)
> +        drawable!!.draw(canvas)
> +    }
> +  }
> +
>
> Nit: Remove extra space
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia-android/pull/188?email_source=notifications&email_token=AD4LXZHNFWPYZUXSAIB6LDDQLV4I5A5CNFSM4I2WFUPKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCGDYYNQ#pullrequestreview-294095926>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD4LXZCR5PUOLMRIUVTK5ALQLV4I5ANCNFSM4I2WFUPA>
> .
>

veena14cs committed:
removed usused layout files

veena14cs commented:
corrected indentation

On Fri, Sep 27, 2019 at 9:04 AM Rajat Talesra <notifications@github.com>
wrote:

> *@rt4914* commented on this pull request.
> ------------------------------
>
> In app/src/main/res/layout/content_interaction_fragment.xml
> <https://github.com/oppia/oppia-android/pull/188#discussion_r328902245>:
>
> > @@ -0,0 +1,14 @@
> +<layout xmlns:android="http://schemas.android.com/apk/res/android">
> +    <LinearLayout
> +      android:id="@+id/rootContainer"
> +      android:layout_width="match_parent"
> +      android:layout_height="match_parent"
> +      android:orientation="vertical">
> +      <RadioGroup
> +        android:id="@+id/radioGroup"
> +        android:layout_width="match_parent"
> +        android:layout_height="wrap_content"
> +      android:padding="15dp">
>
> Indentation issue.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia-android/pull/188?email_source=notifications&email_token=AD4LXZCUXZ3XOW65HYMWBJDQLV5MXA5CNFSM4I2WFUPKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCGDZZOI#pullrequestreview-294100153>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD4LXZA5TYG44NNCU3OPW53QLV5MXANCNFSM4I2WFUPA>
> .
>

rt4914 commented on app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt:
```diff
@@ -0,0 +1,89 @@
+import android.os.Build
+import android.view.Gravity
+
+// TODO(#190): Move this to a custom.inputinteractionview.
```
If here `custom.inputinteractionview` means folders than I guess you can do that right now and it won't create any issue with @nikitamarysolomanpvt 's PR#190

veena14cs replied:
as per our conversations added TODO #163 StateCardFragment Low-fi PR.

rt4914 commented on app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt:
```diff
@@ -0,0 +1,89 @@
+    initViews()
+  }
+
+  //update default attributes of MultipleSelectionInputInteractionView here
```
Please correct this comment format.
`// Update default attributes of MultipleSelectionInputInteractionView here.`

Nit:
There should be space after `//` and `.` at the end of sentence

veena14cs replied:
corrected

rt4914 commented on app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt:
```diff
@@ -0,0 +1,89 @@
+  }
+
+  //update default attributes of MultipleSelectionInputInteractionView here
+  fun initViews() {
```
Nit: There are two types of constructors in kotlin, primary and secondary.

Maybe you need those constructors here.

Referenced: https://blog.mindorks.com/primary-and-secondary-constructors-in-kotlin

rt4914 replied:
Rather than calling `initViews()` inside `init` better implement the entire code of `initViews` inside `init` itself and remove the `initViews method`

veena14cs replied:
added entire code to init()

rt4914 commented on app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt:
```diff
@@ -0,0 +1,89 @@
+
+  var colorStateList = ColorStateList(
+    arrayOf(
+      intArrayOf(android.R.attr.state_enabled) //enabled
```
Remove comment. Your `state_enabled` is self-explanatory.

veena14cs replied:
removed comment

rt4914 commented on app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt:
```diff
@@ -0,0 +1,89 @@
+  )
+
+  fun convertHtmlToString(rawResponse: Any?, rdbtn: View): Spanned {
+
```
Remove this extra space.

we don't add extra space on the first and last line of method because indentation automatically makes everything clear.

veena14cs replied:
ok removed blank space

rt4914 reviewed (commented):
PTAL

rt4914 commented:
https://github.com/oppia/oppia-android/pull/188#issuecomment-535888499

@veena14cs 

i understand that this naming convention is very confusing but it better if we get used to it so that we can communicate easily with @seanlip and @BenHenning , because they are used to these terms whereas we are completely new to this project and it is easier for us to follow this and difficult for them to change this naming now.

veena14cs committed:
fixed const values, fixed constructor

veena14cs committed:
removed extra spaces

veena14cs committed:
Update StateFragmentPresenter.kt

veena14cs committed:
updated todo

rt4914 commented on app/src/main/AndroidManifest.xml:
```diff
@@ -11,8 +11,8 @@
-    <activity android:name=".player.exploration.ExplorationActivity"/>
-    <activity android:name="org.oppia.app.home.HomeActivity">
+    <activity android:name=".player.exploration.ExplorationActivity">
+    <!--<activity android:name="org.oppia.app.home.HomeActivity">-->
```
Revert these changes, we don't need to make sure that the PR code uses your implementation.
Rather than that we want that your PR should be merged directly and whenever the develop runs the develop branch, homefragment should appear.

veena14cs replied:
reverted

rt4914 commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -14,12 +15,12 @@ class ExplorationActivityPresenter @Inject constructor(private val activity: App
-  private fun getExplorationFragment(): ExplorationFragment? {
-    return activity.supportFragmentManager.findFragmentById(R.id.exploration_fragment_placeholder) as ExplorationFragment?
+  private fun getExplorationFragment(): StateFragment? {
+    return activity.supportFragmentManager.findFragmentById(R.id.exploration_fragment_placeholder) as StateFragment?
```
Revert these changes, we don't need to make sure that the PR code uses your implementation.
Rather than that we want that your PR should be merged directly and whenever the develop runs the develop branch, homefragment should appear.

veena14cs replied:
reverted

rt4914 commented on app/src/main/java/org/oppia/app/player/state/CustomRadioButton.kt:
```diff
@@ -0,0 +1,71 @@
+  }
+
+  // Update default attributes of ItemSelectionInputInteractionView here.
+  fun initViews() {
```
Ditto as per earlier comment.

veena14cs replied:
corrected comment.

rt4914 commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -31,4 +43,63 @@ class StateFragmentPresenter @Inject constructor(
   }
+
+  private fun showInputInteractions() {
+
```
As mentioned earlier, we do not need extra black space after opening and closing bracket of any function.

veena14cs replied:
removed blank space

rt4914 commented on app/src/main/res/layout/state_fragment.xml:
```diff
@@ -15,20 +15,36 @@
       app:layout_constraintEnd_toEndOf="parent"
-      app:layout_constraintStart_toStartOf="parent"
-      app:layout_constraintTop_toTopOf="parent">
+      app:layout_constraintStart_toStartOf="parent" app:layout_constraintTop_toTopOf="parent">
```
Correct this code indentation.

rt4914 commented on app/src/main/res/layout/state_fragment.xml:
```diff
@@ -15,20 +15,36 @@
+      android:background="@color/white" android:orientation="vertical"
+      android:padding="10dp"
+      app:layout_constraintBottom_toTopOf="@+id/dummy_audio_button"
+      app:layout_constraintTop_toBottomOf="@+id/linearLayout" tools:layout_editor_absoluteX="8dp">
```
Correct this code indentation

veena14cs replied:
corrected indentation

rt4914 commented on app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentPresenterTest.kt:
```diff
@@ -0,0 +1,101 @@
+//          .perform(click())
+//      onView(ViewMatchers.withHint("text"))
+//          .perform(clearText(), typeText("1/3"));
+//      pauseTestFor(500)
```
Why is this test commented? Also, this is not from your PR, this is from #190

veena14cs replied:
Added it for testing purpose. reverted it.

rt4914 commented on app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentPresenterTest.kt:
```diff
@@ -0,0 +1,101 @@
+
+@RunWith(AndroidJUnit4::class)
+class StateFragmentPresenterTest {
+  //to test text,number,and fraction input interaction views
```
This comment is from PR #190

veena14cs replied:
its been removed

rt4914 commented on utility/src/main/java/org/oppia/util/data/UrlImageParser.kt:
```diff
@@ -0,0 +1,63 @@
+    internal var drawable: Drawable? = null
+
+    override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
+
```
Remove this blank line.

veena14cs replied:
removed

rt4914 commented on utility/src/main/java/org/oppia/util/data/UrlImageParser.kt:
```diff
@@ -0,0 +1,63 @@
+      tvContents.post {
+        val w = tvContents.width
+        val hh = (drawable as BitmapDrawable).intrinsicHeight
+        val ww = (drawable as BitmapDrawable).intrinsicWidth
```
This issue was mentioned earlier too. Rename these variables to something which become self-explanatory.

veena14cs replied:
rename these variables

rt4914 reviewed (commented):
PTAL

veena14cs committed:
removed initviews method

veena14cs committed:
reverted changes

veena14cs committed:
Update UrlImageParser.kt

veena14cs committed:
Update state_fragment.xml

veena14cs committed:
Update StateFragmentPresenterTest.kt

veena14cs commented:
Duplicate of this PR is been created here #[196](https://github.com/oppia/oppia-android/pull/196)
